Add additional heterogeneous overloads.

BenKaufmann opened the pull request:
* Add heterogeneous overloads for try_emplace, insert_or_assign, and operator[] to {b}hopscotch_map and hopscotch_hash.

* Similar overloads were added to the standard associative containers with C++26 (see P2363R5).

Tessil commented:
Thanks, it looks good.

Few comments:
* Would it be possible to also add the changes for the `(b)hopscotch_set`  that are in the P2363R5?
* For the comment with P2363R5, could we use the `This overload only participates in the overload resolution if...` instead for all these new methods?
* At the time heterogeneous overloads were added to the project (2017), only `std::map/set` had these in the standard. The `std::unordered_map/set` ones were added in C++20. 
I used the same convention as `std::map`, requiring `has_is_transparent<KeyEqual>` but I didn't check for `has_is_transparent<Compare>`.
Ideally this should be changed but I have kept it like that for now to keep backward compatibility.
Could the new methods also only checks `has_is_transparent<KeyEqual>` for consistency?

Thanks

Tessil commented:
Thank you for the change and sorry for the delay.

> I don't think that we can remove the is_convertible check from the try_emplace overloads as otherwise,
try_emplace(K&& k, Args&&... args) and try_emplace(const_iterator hint, K&& k, Args&&... args) are ambiguous. However, I adjusted the code such that it no longer uses std::is_convertible_v.

Sorry, I meant to only have `has_is_transparent<KeyEqual>` as is transparent check and not `has_is_transparent<Compare>` but keep all other necessary checks  (`is_convertible`). Could you check to put them back? Thanks!

I just want to avoid incoherences in the API were some heterogeneous methods require only `has_is_transparent<KeyEqual>` but other require `has_is_transparent<KeyEqual> & has_is_transparent<Compare>`.

BenKaufmann commented:
> Sorry, I meant to only have `has_is_transparent<KeyEqual>` as is transparent check and not `has_is_transparent<Compare>` but keep all other necessary checks (`is_convertible`). Could you check to put them back? Thanks!
> 
> I just want to avoid incoherences in the API were some heterogeneous methods require only `has_is_transparent<KeyEqual>` but other require `has_is_transparent<KeyEqual> & has_is_transparent<Compare>`.

Sorry, I'm not sure I can follow:

- I did not remove the `is_convertible` checks. I only changed my previous use of `std::is_convertible_v` (added with C++17) to a C++11 compatible `std::is_convertible<...>::value` under the assumption that this library should work with a C++11 standard-compliant compiler.
- I added `has_is_transparent<Compare>` in some places precisely to avoid further incoherences. Both `bhopscotch_map.h` and `bhopscotch_set.h` already require `has_is_transparent<Compare>` in the existing heterogeneous functions, like .e.g. `erase()`. So, for consistency, I also added this requirement to the new functions. 

Could you please clarify your concerns so that I can update this PR again if needed?

Tessil commented:
Sorry if I was unclear, it seems some `std::is_convertible` were removed ([example](https://github.com/Tessil/hopscotch-map/pull/63/commits/6497853574987e4b3599da3d65ebcd2a0bbbe09f#diff-1957abbc110fad9c1926dc31e7d08987d01d8516f905921c03e8c3f17fb4f63aR341)) even though they are part of P2363R5. Is there any reason why they were removed?

BenKaufmann commented:
> it seems some std::is_convertible were removed [...]  Is there any reason why they were removed?

Given that `iterator` is convertible to `const_iterator`, I assumed that checking for the latter is enough. Of course, this assumption is wrong in the general case. I now restored the previously removed checks and also added a test for the constraints.

Tessil commented:
Sorry for the review delay and thanks for your patience.

It looks good to me in general. Just a bit worried about the `static int inst` in `heterogeneous_test` if we ever want to run the tests in parallel, but it isn't supported by Boost.Test currently anyway. It can always be changed later.

I merged the change.

Thanks again for your contribution.

BenKaufmann commented:
> Just a bit worried about the `static int inst` in `heterogeneous_test` if we ever want to run the tests in parallel, but it isn't supported by Boost.Test currently anyway. It can always be changed later.

I understand your concern and drafted an alternative approach: https://github.com/Tessil/hopscotch-map/compare/master...BenKaufmann:hopscotch-map:simplify-unit-tests.
